{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e00e44c8-f1b5-44c4-ad98-c19a95f548b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                           Version\n",
      "--------------------------------- ------------\n",
      "absl-py                           2.1.0\n",
      "aiobotocore                       2.7.0\n",
      "aiohttp                           3.9.3\n",
      "aioitertools                      0.7.1\n",
      "aiosignal                         1.2.0\n",
      "alabaster                         0.7.12\n",
      "altair                            5.0.1\n",
      "anaconda-anon-usage               0.4.3\n",
      "anaconda-catalogs                 0.2.0\n",
      "anaconda-client                   1.12.3\n",
      "anaconda-cloud-auth               0.1.4\n",
      "anaconda-navigator                2.5.2\n",
      "anaconda-project                  0.11.1\n",
      "anyio                             4.2.0\n",
      "appdirs                           1.4.4\n",
      "archspec                          0.2.1\n",
      "argon2-cffi                       21.3.0\n",
      "argon2-cffi-bindings              21.2.0\n",
      "arrow                             1.2.3\n",
      "astroid                           2.14.2\n",
      "astropy                           5.3.4\n",
      "asttokens                         2.0.5\n",
      "astunparse                        1.6.3\n",
      "async-lru                         2.0.4\n",
      "atomicwrites                      1.4.0\n",
      "attrs                             23.1.0\n",
      "Automat                           20.2.0\n",
      "autopep8                          1.6.0\n",
      "Babel                             2.11.0\n",
      "backports.functools-lru-cache     1.6.4\n",
      "backports.tempfile                1.0\n",
      "backports.weakref                 1.0.post1\n",
      "bcrypt                            3.2.0\n",
      "beautifulsoup4                    4.12.2\n",
      "binaryornot                       0.4.4\n",
      "black                             23.11.0\n",
      "bleach                            4.1.0\n",
      "blinker                           1.6.2\n",
      "bokeh                             3.3.4\n",
      "boltons                           23.0.0\n",
      "botocore                          1.31.64\n",
      "Bottleneck                        1.3.7\n",
      "Brotli                            1.0.9\n",
      "cachetools                        4.2.2\n",
      "certifi                           2024.2.2\n",
      "cffi                              1.16.0\n",
      "chardet                           4.0.0\n",
      "charset-normalizer                2.0.4\n",
      "click                             8.1.7\n",
      "cloudpickle                       2.2.1\n",
      "clyent                            1.2.2\n",
      "colorama                          0.4.6\n",
      "colorcet                          3.0.1\n",
      "comm                              0.1.2\n",
      "conda                             24.1.2\n",
      "conda-build                       24.1.2\n",
      "conda-content-trust               0.2.0\n",
      "conda_index                       0.4.0\n",
      "conda-libmamba-solver             24.1.0\n",
      "conda-pack                        0.6.0\n",
      "conda-package-handling            2.2.0\n",
      "conda_package_streaming           0.9.0\n",
      "conda-repo-cli                    1.0.75\n",
      "conda-token                       0.4.0\n",
      "conda-verify                      3.4.2\n",
      "constantly                        23.10.4\n",
      "contourpy                         1.2.0\n",
      "cookiecutter                      2.5.0\n",
      "cryptography                      42.0.2\n",
      "cssselect                         1.2.0\n",
      "cycler                            0.11.0\n",
      "cytoolz                           0.12.2\n",
      "dask                              2023.11.0\n",
      "datashader                        0.16.0\n",
      "debugpy                           1.6.7\n",
      "decorator                         5.1.1\n",
      "defusedxml                        0.7.1\n",
      "diff-match-patch                  20200713\n",
      "dill                              0.3.7\n",
      "distributed                       2023.11.0\n",
      "distro                            1.8.0\n",
      "docstring-to-markdown             0.11\n",
      "docutils                          0.18.1\n",
      "entrypoints                       0.4\n",
      "et-xmlfile                        1.1.0\n",
      "executing                         0.8.3\n",
      "fastjsonschema                    2.16.2\n",
      "filelock                          3.13.1\n",
      "flake8                            6.0.0\n",
      "Flask                             2.2.5\n",
      "flatbuffers                       24.3.25\n",
      "fonttools                         4.25.0\n",
      "frozenlist                        1.4.0\n",
      "fsspec                            2023.10.0\n",
      "future                            0.18.3\n",
      "gast                              0.6.0\n",
      "gensim                            4.3.0\n",
      "gitdb                             4.0.7\n",
      "GitPython                         3.1.37\n",
      "gmpy2                             2.1.2\n",
      "google-pasta                      0.2.0\n",
      "greenlet                          3.0.1\n",
      "grpcio                            1.65.4\n",
      "h5py                              3.11.0\n",
      "HeapDict                          1.0.1\n",
      "holoviews                         1.18.3\n",
      "hvplot                            0.9.2\n",
      "hyperlink                         21.0.0\n",
      "idna                              3.4\n",
      "imagecodecs                       2023.1.23\n",
      "imageio                           2.33.1\n",
      "imagesize                         1.4.1\n",
      "imbalanced-learn                  0.11.0\n",
      "importlib-metadata                7.0.1\n",
      "incremental                       22.10.0\n",
      "inflection                        0.5.1\n",
      "iniconfig                         1.1.1\n",
      "intake                            0.6.8\n",
      "intel-tensorflow                  0.0.1\n",
      "intervaltree                      3.1.0\n",
      "ipykernel                         6.28.0\n",
      "ipython                           8.20.0\n",
      "ipython-genutils                  0.2.0\n",
      "ipywidgets                        7.6.5\n",
      "isort                             5.9.3\n",
      "itemadapter                       0.3.0\n",
      "itemloaders                       1.1.0\n",
      "itsdangerous                      2.0.1\n",
      "jaraco.classes                    3.2.1\n",
      "jedi                              0.18.1\n",
      "jellyfish                         1.0.1\n",
      "Jinja2                            3.1.3\n",
      "jmespath                          1.0.1\n",
      "joblib                            1.2.0\n",
      "json5                             0.9.6\n",
      "jsonpatch                         1.32\n",
      "jsonpointer                       2.1\n",
      "jsonschema                        4.19.2\n",
      "jsonschema-specifications         2023.7.1\n",
      "jupyter                           1.0.0\n",
      "jupyter_client                    8.6.0\n",
      "jupyter-console                   6.6.3\n",
      "jupyter_core                      5.5.0\n",
      "jupyter-events                    0.8.0\n",
      "jupyter-lsp                       2.2.0\n",
      "jupyter_server                    2.10.0\n",
      "jupyter_server_terminals          0.4.4\n",
      "jupyterlab                        4.0.11\n",
      "jupyterlab-pygments               0.1.2\n",
      "jupyterlab_server                 2.25.1\n",
      "jupyterlab-widgets                3.0.9\n",
      "keras                             3.4.1\n",
      "keyring                           23.13.1\n",
      "kiwisolver                        1.4.4\n",
      "lazy_loader                       0.3\n",
      "lazy-object-proxy                 1.6.0\n",
      "lckr_jupyterlab_variableinspector 3.1.0\n",
      "libarchive-c                      2.9\n",
      "libclang                          18.1.1\n",
      "libmambapy                        1.5.6\n",
      "linkify-it-py                     2.0.0\n",
      "llvmlite                          0.42.0\n",
      "lmdb                              1.4.1\n",
      "locket                            1.0.0\n",
      "lxml                              5.3.0\n",
      "lz4                               4.3.2\n",
      "Markdown                          3.4.1\n",
      "markdown-it-py                    2.2.0\n",
      "MarkupSafe                        2.1.3\n",
      "matplotlib                        3.8.0\n",
      "matplotlib-inline                 0.1.6\n",
      "mccabe                            0.7.0\n",
      "mdit-py-plugins                   0.3.0\n",
      "mdurl                             0.1.0\n",
      "menuinst                          2.0.2\n",
      "mistune                           2.0.4\n",
      "mkl-fft                           1.3.8\n",
      "mkl-random                        1.2.4\n",
      "mkl-service                       2.4.0\n",
      "ml-dtypes                         0.4.0\n",
      "more-itertools                    10.1.0\n",
      "mpmath                            1.3.0\n",
      "msgpack                           1.0.3\n",
      "multidict                         6.0.4\n",
      "multipledispatch                  0.6.0\n",
      "munkres                           1.1.4\n",
      "mypy                              1.8.0\n",
      "mypy-extensions                   1.0.0\n",
      "namex                             0.0.8\n",
      "navigator-updater                 0.4.0\n",
      "nbclient                          0.8.0\n",
      "nbconvert                         7.10.0\n",
      "nbformat                          5.9.2\n",
      "nest-asyncio                      1.6.0\n",
      "networkx                          3.1\n",
      "nltk                              3.8.1\n",
      "notebook                          7.0.8\n",
      "notebook_shim                     0.2.3\n",
      "numba                             0.59.0\n",
      "numexpr                           2.8.7\n",
      "numpy                             1.26.4\n",
      "numpydoc                          1.5.0\n",
      "opencv-python                     4.10.0.84\n",
      "openpyxl                          3.0.10\n",
      "opt-einsum                        3.3.0\n",
      "optree                            0.12.1\n",
      "overrides                         7.4.0\n",
      "packaging                         23.1\n",
      "pandas                            2.1.4\n",
      "pandocfilters                     1.5.0\n",
      "panel                             1.3.8\n",
      "param                             2.0.2\n",
      "paramiko                          2.8.1\n",
      "parsel                            1.8.1\n",
      "parso                             0.8.3\n",
      "partd                             1.4.1\n",
      "pathlib                           1.0.1\n",
      "pathspec                          0.10.3\n",
      "patsy                             0.5.3\n",
      "pexpect                           4.8.0\n",
      "pickleshare                       0.7.5\n",
      "pillow                            10.2.0\n",
      "pip                               24.2\n",
      "pkce                              1.0.3\n",
      "pkginfo                           1.9.6\n",
      "platformdirs                      3.10.0\n",
      "plotly                            5.9.0\n",
      "pluggy                            1.0.0\n",
      "ply                               3.11\n",
      "prometheus-client                 0.14.1\n",
      "prompt-toolkit                    3.0.43\n",
      "Protego                           0.1.16\n",
      "protobuf                          3.20.3\n",
      "psutil                            5.9.0\n",
      "ptyprocess                        0.7.0\n",
      "pure-eval                         0.2.2\n",
      "py-cpuinfo                        9.0.0\n",
      "pyarrow                           14.0.2\n",
      "pyasn1                            0.4.8\n",
      "pyasn1-modules                    0.2.8\n",
      "pycodestyle                       2.10.0\n",
      "pycosat                           0.6.6\n",
      "pycparser                         2.21\n",
      "pyct                              0.5.0\n",
      "pycurl                            7.45.2\n",
      "pydantic                          1.10.12\n",
      "pydeck                            0.8.0\n",
      "PyDispatcher                      2.0.5\n",
      "pydocstyle                        6.3.0\n",
      "pyerfa                            2.0.0\n",
      "pyflakes                          3.0.1\n",
      "Pygments                          2.15.1\n",
      "PyJWT                             2.4.0\n",
      "pylint                            2.16.2\n",
      "pylint-venv                       2.3.0\n",
      "pyls-spyder                       0.4.0\n",
      "PyNaCl                            1.5.0\n",
      "pyodbc                            5.0.1\n",
      "pyOpenSSL                         24.0.0\n",
      "pyparsing                         3.0.9\n",
      "PyQt5                             5.15.11\n",
      "PyQt5-Qt5                         5.15.2\n",
      "PyQt5_sip                         12.15.0\n",
      "PyQtWebEngine                     5.15.6\n",
      "PySocks                           1.7.1\n",
      "pytest                            7.4.0\n",
      "python-dateutil                   2.8.2\n",
      "python-dotenv                     0.21.0\n",
      "python-json-logger                2.0.7\n",
      "python-lsp-black                  1.2.1\n",
      "python-lsp-jsonrpc                1.0.0\n",
      "python-lsp-server                 1.7.2\n",
      "python-slugify                    5.0.2\n",
      "python-snappy                     0.6.1\n",
      "pytoolconfig                      1.2.6\n",
      "pytz                              2023.3.post1\n",
      "pyviz_comms                       3.0.0\n",
      "pywavelets                        1.5.0\n",
      "pywin32                           305.1\n",
      "pywin32-ctypes                    0.2.0\n",
      "pywinpty                          2.0.10\n",
      "PyYAML                            6.0.1\n",
      "pyzmq                             25.1.2\n",
      "QDarkStyle                        3.0.2\n",
      "qstylizer                         0.2.2\n",
      "QtAwesome                         1.2.2\n",
      "qtconsole                         5.4.2\n",
      "QtPy                              2.4.1\n",
      "queuelib                          1.6.2\n",
      "referencing                       0.30.2\n",
      "regex                             2023.10.3\n",
      "requests                          2.31.0\n",
      "requests-file                     1.5.1\n",
      "requests-toolbelt                 1.0.0\n",
      "rfc3339-validator                 0.1.4\n",
      "rfc3986-validator                 0.1.1\n",
      "rich                              13.3.5\n",
      "rope                              1.7.0\n",
      "rpds-py                           0.10.6\n",
      "Rtree                             1.0.1\n",
      "ruamel.yaml                       0.17.21\n",
      "ruamel-yaml-conda                 0.17.21\n",
      "s3fs                              2023.10.0\n",
      "scikit-image                      0.22.0\n",
      "scikit-learn                      1.2.2\n",
      "scipy                             1.11.4\n",
      "Scrapy                            2.8.0\n",
      "seaborn                           0.12.2\n",
      "semver                            2.13.0\n",
      "Send2Trash                        1.8.2\n",
      "service-identity                  18.1.0\n",
      "setuptools                        68.2.2\n",
      "sip                               6.7.12\n",
      "six                               1.16.0\n",
      "smart-open                        5.2.1\n",
      "smmap                             4.0.0\n",
      "sniffio                           1.3.0\n",
      "snowballstemmer                   2.2.0\n",
      "sortedcontainers                  2.4.0\n",
      "soupsieve                         2.5\n",
      "Sphinx                            5.0.2\n",
      "sphinxcontrib-applehelp           1.0.2\n",
      "sphinxcontrib-devhelp             1.0.2\n",
      "sphinxcontrib-htmlhelp            2.0.0\n",
      "sphinxcontrib-jsmath              1.0.1\n",
      "sphinxcontrib-qthelp              1.0.3\n",
      "sphinxcontrib-serializinghtml     1.1.5\n",
      "spyder                            5.4.3\n",
      "spyder-kernels                    2.4.4\n",
      "SQLAlchemy                        2.0.25\n",
      "stack-data                        0.2.0\n",
      "statsmodels                       0.14.0\n",
      "streamlit                         1.30.0\n",
      "sympy                             1.12\n",
      "tables                            3.9.2\n",
      "tabulate                          0.9.0\n",
      "tblib                             1.7.0\n",
      "tenacity                          8.2.2\n",
      "tensorboard                       2.17.0\n",
      "tensorboard-data-server           0.7.2\n",
      "tensorflow                        2.17.0\n",
      "tensorflow-intel                  2.17.0\n",
      "tensorflow-io-gcs-filesystem      0.31.0\n",
      "termcolor                         2.4.0\n",
      "terminado                         0.17.1\n",
      "text-unidecode                    1.3\n",
      "textdistance                      4.2.1\n",
      "threadpoolctl                     2.2.0\n",
      "three-merge                       0.1.1\n",
      "tifffile                          2023.4.12\n",
      "tinycss2                          1.2.1\n",
      "tldextract                        3.2.0\n",
      "toml                              0.10.2\n",
      "tomlkit                           0.11.1\n",
      "toolz                             0.12.0\n",
      "tornado                           6.3.3\n",
      "tqdm                              4.65.0\n",
      "traitlets                         5.7.1\n",
      "truststore                        0.8.0\n",
      "Twisted                           23.10.0\n",
      "twisted-iocpsupport               1.0.2\n",
      "typing_extensions                 4.9.0\n",
      "tzdata                            2023.3\n",
      "tzlocal                           2.1\n",
      "uc-micro-py                       1.0.1\n",
      "ujson                             5.4.0\n",
      "Unidecode                         1.2.0\n",
      "urllib3                           2.0.7\n",
      "validators                        0.18.2\n",
      "w3lib                             2.1.2\n",
      "watchdog                          2.1.6\n",
      "wcwidth                           0.2.5\n",
      "webencodings                      0.5.1\n",
      "websocket-client                  0.58.0\n",
      "Werkzeug                          2.2.3\n",
      "wget                              3.2\n",
      "whatthepatch                      1.0.2\n",
      "wheel                             0.41.2\n",
      "widgetsnbextension                3.5.2\n",
      "win-inet-pton                     1.1.0\n",
      "wrapt                             1.14.1\n",
      "xarray                            2023.6.0\n",
      "xlwings                           0.29.1\n",
      "xyzservices                       2022.9.0\n",
      "yapf                              0.31.0\n",
      "yarl                              1.9.3\n",
      "zict                              3.0.0\n",
      "zipp                              3.17.0\n",
      "zope.interface                    5.4.0\n",
      "zstandard                         0.19.0\n"
     ]
    }
   ],
   "source": [
    "!pip list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0db32c9c-8598-47da-9f89-a48ff319d1ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: networkx in c:\\programdata\\anaconda3\\lib\\site-packages (3.1)\n",
      "Requirement already satisfied: matplotlib in c:\\programdata\\anaconda3\\lib\\site-packages (3.8.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Could not find a version that satisfies the requirement tkinter (from versions: none)\n",
      "ERROR: No matching distribution found for tkinter\n"
     ]
    }
   ],
   "source": [
    "pip install networkx matplotlib tkinter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "50371c19-3511-4b46-bb40-7435f574161f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8.6\n"
     ]
    }
   ],
   "source": [
    "import tkinter as tk\n",
    "print(tk.TkVersion)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "593f4cde-9cb1-4005-a8ec-0da1ba9f7acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7e024685-fe54-4e84-9238-21450e7b9c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "G = nx.Graph()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3e56f636-4be7-42b2-8e46-7ed72a4f95f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "G.add_edge('A', 'B', weight=5)\n",
    "G.add_edge('B', 'C', weight=3)\n",
    "G.add_edge('A', 'C', weight=9)\n",
    "G.add_edge('C', 'D', weight=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bd58d11c-3592-49ef-a236-df5fe9eaaa5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_traffic(G):\n",
    "    for u, v, data in G.edges(data=True):\n",
    "        data['weight'] = random.randint(1, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3ef45c41-6093-4166-90e0-bd682fed7c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "update_traffic(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4aec2940-d29a-4c24-8477-4bc1caeeb983",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_shortest_path(G, source, target):\n",
    "    return nx.dijkstra_path(G, source, target, weight='weight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3cb17df8-8ce5-44c3-adce-8fcacf658e55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shortest path: ['A', 'C', 'D']\n"
     ]
    }
   ],
   "source": [
    "path = find_shortest_path(G, 'A', 'D')\n",
    "print(\"Shortest path:\", path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a16dcb89-7120-497a-8058-6c666cc0a9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "from tkinter import messagebox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4e5d459e-6a59-4799-a779-e5e6490076fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_shortest_path():\n",
    "    source = entry_source.get()\n",
    "    target = entry_target.get()\n",
    "    path = find_shortest_path(G, source, target)\n",
    "    messagebox.showinfo(\"Shortest Path\", f\"Shortest path from {source} to {target}: {path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4e74427d-eb0b-4b59-bfc2-956392aad783",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root = tk.Tk()\n",
    "root.title(\"Traffic System\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5a34cfba-2dfb-4e75-b701-46fc8132d7ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk.Label(root, text=\"Source:\").grid(row=0, column=0)\n",
    "entry_source = tk.Entry(root)\n",
    "entry_source.grid(row=0, column=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d66afc4b-4af2-45d7-9bae-9b2afe7c4483",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk.Label(root, text=\"Target:\").grid(row=1, column=0)\n",
    "entry_target = tk.Entry(root)\n",
    "entry_target.grid(row=1, column=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "83cc7d75-e405-4b5c-ae24-00378168cbb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk.Button(root, text=\"Find Shortest Path\", command=show_shortest_path).grid(row=2, column=0, columnspan=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa386393-c3f7-46f4-bbcb-829923984ca9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A,C,D not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A,C,D not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A,C,D,B not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A,C,D,B not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A,C,D,B not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A(1),B(2),C(5) not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A(1),B(2),C(5) not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A(1) not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 743, in multi_source_dijkstra\n",
      "    raise nx.NodeNotFound(f\"Node {s} not found in graph\")\n",
      "networkx.exception.NodeNotFound: Node A(1) not found in graph\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 754, in multi_source_dijkstra\n",
      "    return (dist[target], paths[target])\n",
      "            ~~~~^^^^^^^^\n",
      "KeyError: 'E'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 756, in multi_source_dijkstra\n",
      "    raise nx.NetworkXNoPath(f\"No path to {target}.\") from err\n",
      "networkx.exception.NetworkXNoPath: No path to E.\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 754, in multi_source_dijkstra\n",
      "    return (dist[target], paths[target])\n",
      "            ~~~~^^^^^^^^\n",
      "KeyError: 'E'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 756, in multi_source_dijkstra\n",
      "    raise nx.NetworkXNoPath(f\"No path to {target}.\") from err\n",
      "networkx.exception.NetworkXNoPath: No path to E.\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 754, in multi_source_dijkstra\n",
      "    return (dist[target], paths[target])\n",
      "            ~~~~^^^^^^^^\n",
      "KeyError: 'E'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 756, in multi_source_dijkstra\n",
      "    raise nx.NetworkXNoPath(f\"No path to {target}.\") from err\n",
      "networkx.exception.NetworkXNoPath: No path to E.\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 754, in multi_source_dijkstra\n",
      "    return (dist[target], paths[target])\n",
      "            ~~~~^^^^^^^^\n",
      "KeyError: 'E'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\2449603955.py\", line 4, in show_shortest_path\n",
      "    path = find_shortest_path(G, source, target)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\dell\\AppData\\Local\\Temp\\ipykernel_4172\\1851588410.py\", line 2, in find_shortest_path\n",
      "    return nx.dijkstra_path(G, source, target, weight='weight')\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 167, in dijkstra_path\n",
      "    (length, path) = single_source_dijkstra(G, source, target=target, weight=weight)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 481, in single_source_dijkstra\n",
      "    return multi_source_dijkstra(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\ProgramData\\anaconda3\\Lib\\site-packages\\networkx\\algorithms\\shortest_paths\\weighted.py\", line 756, in multi_source_dijkstra\n",
      "    raise nx.NetworkXNoPath(f\"No path to {target}.\") from err\n",
      "networkx.exception.NetworkXNoPath: No path to E.\n"
     ]
    }
   ],
   "source": [
    "root.mainloop()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fadf09eb-f8e3-4b5f-ade6-945b9498be44",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_and_refresh():\n",
    "    update_traffic(G)\n",
    "    root.after(5000, update_and_refresh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "04b04db9-7109-43a0-8dff-4d72dcb869f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "update_and_refresh()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3ecef63f-e318-412d-a9b3-245824383b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "root.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aababd8-f943-46f8-b487-96ac9a003160",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
